### PR TITLE
add a default to `ResultData`, some related cleanup

### PR DESCRIPTION
**Pull request by samuelcolvin**

### CODE REVIEW

File: pydantic_ai_slim/pydantic_ai/result.py [hunk 1]
```diff
@@ -4,39 +4,56 @@
...
 """Type variable for the result data of a run."""
 
+ResultValidatorFunc = Union[
+    Callable[[RunContext[AgentDeps], ResultData], ResultData],
+    Callable[[RunContext[AgentDeps], ResultData], Awaitable[ResultData]],
+    Callable[[ResultData], ResultData],
+    Callable[[ResultData], Awaitable[ResultData]],
+]
+"""
+A function that always takes `ResultData` and returns `ResultData` but:
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/result.py]:
```suggestion
A function that always takes `ResultData` and returns `ResultData` and:
```

File: pydantic_ai_slim/pydantic_ai/result.py [hunk 2]
```diff
@@ -4,39 +4,56 @@
...
+
+* may or maybe not take [`RunContext`][pydantic_ai.tools.RunContext] as a first argument
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/result.py]:
```suggestion
* may or may not take [`RunContext`][pydantic_ai.tools.RunContext] as a first argument
```

File: pydantic_ai_slim/pydantic_ai/result.py [hunk 3]
```diff
@@ -4,39 +4,56 @@
...
+* and, may or may not be async
```
Comment (sydney-runkle) [inline pydantic_ai_slim/pydantic_ai/result.py]:
```suggestion
* may or may not be async
```

File: tests/typed_agent.py [hunk 4]
```diff
@@ -178,6 +178,13 @@ def run_sync3() -> None:
     assert_type(result.data, Union[Foo, Bar])
 
 
+MyUnion: TypeAlias = 'Foo | Bar'
+union_agent2: Agent[None, MyUnion] = Agent(
+    result_type=MyUnion,  # type: ignore[arg-type]
```
Comment (sydney-runkle) [inline tests/typed_agent.py:183]: Why do we need the type ignore here? What's the error?

**Review by sydney-runkle — COMMENTED** [review]
Looking good, left a few nit picky comments 👍

File: tests/typed_agent.py (hunk 4 above)
Comment (samuelcolvin) [reply tests/typed_agent.py:183]: we need the type here so we can enforce the return type. The error is that a union is not a type, but a typeform, docs about waiting for PEP 747 - https://ai.pydantic.dev/results/?h=pep#structured-result-validation

**Pull request merged**